### PR TITLE
Add rnfFoldableLTR and rnfFoldableRTL

### LINKED ISSUE #17: Offer Foldable helper
It's possible to define a reasonable `NFData` instance for any `Foldable` type. We can offer a suitable `rnf` to help.

```haskell
data Unit = Unit

instance Monoid Unit where
  mempty = Unit
  Unit `mappend` Unit = Unit -- strict in both arguments, unlike ()

rnfFoldable :: (Foldable f, NFData a) => f a -> ()
rnfFoldable xs = foldMap (\x -> rnf x `seq` Unit) xs `seq` ()
```

### PR DESCRIPTION
**Pull request by treeowl**
If a type has a `Foldable` instance, then we can force it from
left to right or from right to left.

Fixes #17

### CODE REVIEW

**Comment by RyanGlScott** [comment]
For context: Haskell libraries mailing list discussion: https://mail.haskell.org/pipermail/libraries/2016-July/027164.html

**Comment by rwbarton** [comment]
This should carry a huge warning that it is normally not a sensible definition for `Foldable` instances where `f a` contains fields of types other than `a`, such as a pair.

**Comment by treeowl** [comment]
It has such a warning, specifically mentioning `Either a` and `(,) a`. Is
the warning insufficient?

**Comment by rwbarton** [comment]
Oh, so it does. There is such a lot of noise in the diff that I missed it.

File: Control/DeepSeq.hs [hunk 1]
```diff
@@ -202,6 +215,87 @@ f $!! x = x `deepseq` f x
...
+newtype UnitLTR = UnitLTR {getUnitLTR :: ()}
+
+#if MIN_VERSION_base(4,9,0)
+instance Semi.Semigroup UnitLTR where
+  UnitLTR () <> y = y
+#endif
+
+instance Monoid UnitLTR where
+  mempty = UnitLTR ()
+  UnitLTR () `mappend` y = y
```
Comment (RyanGlScott) [inline Control/DeepSeq.hs]: To avoid invoking the wrath of `-Wnoncanonical-monoid-instances` in later GHCs, it'd be prudent to define `mappend` like this:

```haskell
instance Monoid UnitLTR where
  mempty = UnitLTR ()
#if MIN_VERSION_base(4,9,0)
  mappend = (<>)
#else
  UnitLTR () `mappend` y = y
#endif
```

And similarly for `UnitRTL`.

File: Control/DeepSeq.hs [hunk 2]
```diff
@@ -5,16 +5,24 @@
...
 {-# LANGUAGE Safe #-}
+#endif
 # endif
 #endif
 #if __GLASGOW_HASKELL__ >= 706
 {-# LANGUAGE PolyKinds #-}
 #endif
+#if __GLASGOW_HASKELL__ >= 708
+{-# LANGUAGE ScopedTypeVariables #-}
+#endif
```
Comment (RyanGlScott) [inline Control/DeepSeq.hs]: I'd prefer to enable `ScopedTypeVariables` unconditionally (without an `#ifdef`), since all versions of GHC that we test against support `ScopedTypeVariables` anyway.

File: deepseq.cabal [hunk 3]
```diff
@@ -53,6 +53,11 @@ library
     if impl(ghc < 7.6)
       build-depends: ghc-prim == 0.2.*
 
+  if impl(ghc >= 7.8)
+    other-extensions:
+      ScopedTypeVariables
+      Trustworthy
```
Comment (RyanGlScott) [inline deepseq.cabal]: Why are these here if you enable them in `Control.DeepSeq` anyway?

File: Control/DeepSeq.hs [hunk 4]
```diff
@@ -105,6 +114,10 @@ import GHC.Stack ( CallStack(..) )
 import GHC.SrcLoc ( SrcLoc(..) )
 #endif
 
+#if __GLASGOW_HASKELL__ >= 708
```
Comment (RyanGlScott) [inline Control/DeepSeq.hs]: The `#if __GLASGOW_HASKELL__ >= 708` checks should be changed to `#if MIN_VERSION_base(4,7,0)`, since it's just guarding the use of `Data.Coerce.coerce` from `base`.

Comment (treeowl) [inline Control/DeepSeq.hs]: I disagree. The whole `Coercible` apparatus is very much GHC-only. I don't think pretending this is about `base` does anyone any favors.

File: Control/DeepSeq.hs (hunk 2 above)
Comment (treeowl) [inline Control/DeepSeq.hs]: That's fine.

File: Control/DeepSeq.hs (hunk 1 above)
Comment (treeowl) [inline Control/DeepSeq.hs]: Sure.

File: deepseq.cabal (hunk 3 above)
Comment (treeowl) [inline deepseq.cabal]: It's not really important. `other-extensions` is only informative; it doesn't actually enable anything. That's just saying that some modules use those extensions with those versions. I can step them out for sure.

File: Control/DeepSeq.hs (hunk 4 above)
Comment (RyanGlScott) [inline Control/DeepSeq.hs]: Well, it's not something that personally bothers me that much, but I know @hvr has been wanting to be able to decouple `base` installations from GHC versions for a while, so I don't know if tying an conditional import from `base` to a GHC version will make things shaky in the future...

File: deepseq.cabal (hunk 3 above)
Comment (RyanGlScott) [inline deepseq.cabal]: Oops, I misread the phrase `other-extensions`.

Hm, the current full list of `other-extensions` is a bit of a lie since some extensions are only enabled conditionally (e.g., `DeriveGeneric`), but since it's only informative, that doesn't bother me too much. You could probably just merge these two entries with the others for simplicity's sake.

Comment (RyanGlScott) [inline deepseq.cabal]: Ack, I misread again. I was looking at the `other-extensions` for the test suite, not the library, so my above comment doesn't really make sense.

So it's really only the placement of `ScopedTypeVariables` in this list that would need to change. Sorry for the confusion.

Comment (hvr) [inline deepseq.cabal]: @RyanGlScott actually, `other-extensions` is more than just informal. Cabal's solver (since 1.24) takes that list into account when solving install plans and makes sure that the current compiler supports all `other-extensions`.

File: Control/DeepSeq.hs (hunk 4 above)
Comment (hvr) [inline Control/DeepSeq.hs]: Indeed, in general I dislike `__GLASGOW_HASKELL__` being abused as a proxy for `MIN_VERSION_base`.

In this case, however, it's not so clear cut IMHO, as `Coercible` is both a `base`-feature  as well as depending on facilities provided only by certain GHC versions (although I'm not sure if you couldn't actually fake/emulate `Data.Coerce.coerce` via `unsafeCoerce`... in which case I would see this more as a case for `MIN_VERSION_base`)

**Comment by chessai** [comment]
This seems useful, can this be worked in to the modern deepseq?

**Comment by cartazio** [comment]
Perhaps I’m overlooking something obvious : but can’t we tease out
assoxiativity of a structure from its foldmap?

**Comment by chessai** [comment]
Are we worried at all about the loss of `Safe`?

**Comment by mixphix** [comment]
`foldMap` can be more efficient than either `foldl` or `foldr`, so `Unit` was added by #106 (same as UnitLTR).
